packages/core/test/actor.test.ts

```diff
@@ -1105,4 +1105,79 @@ describe('actors', () => {
       })
       .start();
   });
+
+  // https://github.com/statelyai/xstate/issues/2507
+  it('should not crash on sync child completion before stopping', () => {
+    const childMachine = createMachine({
+      initial: 'idle',
+      states: {
+        idle: {
+          always: [
+            {
+              target: 'stopped'
+            }
+          ]
+        },
+        stopped: {
+          type: 'final'
+        }
+      }
+    });
+
+    const parentMachine = createMachine<{
+      child: ActorRefFrom<typeof childMachine> | null;
+    }>(
+      {
+        context: {
+          child: null
+        },
+        entry: 'setup'
+      },
+      {
+        actions: {
+          setup: assign({
+            child: () => spawn(childMachine)
+          })
+        }
+      }
+    );
+    const service = interpret(parentMachine);
+    expect(() => {
+      service.start();
+    }).not.toThrow();
+  });
+
+  it('should not crash on immediate spawned promise completion', () => {
+    const parentMachine = createMachine<{
+      child: ActorRef<any> | null;
+    }>({
+      context: {
+        child: null
+      },
+      entry: assign({
+        child: () => spawn(new Promise((res) => res(null)))
+      })
+    });
+    const service = interpret(parentMachine);
+    expect(() => {
+      service.start();
+    }).not.toThrow();
+  });
+
+  it('should not crash on immediate spawned observable completion', () => {
+    const parentMachine = createMachine<{
+      child: ActorRef<any> | null;
+    }>({
+      context: {
+        child: null
+      },
+      entry: assign({
+        child: () => spawn(EMPTY)
+      })
+    });
+    const service = interpret(parentMachine);
+    expect(() => {
+      service.start();
+    }).not.toThrow();
+  });
```

Comment:
This actually still throws in the `removeChild` method - it's just not obvious due to how RxJS handles errors.

They wrap handlers:
https://github.com/ReactiveX/rxjs/blob/3d062a2b8c2e49efc92e1205ae926c7182745b8c/src/internal/Subscriber.ts#L179
catch them and report:
https://github.com/ReactiveX/rxjs/blob/3d062a2b8c2e49efc92e1205ae926c7182745b8c/src/internal/Subscriber.ts#L195-L202
where reporting is done in a timeout:
https://github.com/ReactiveX/rxjs/blob/3d062a2b8c2e49efc92e1205ae926c7182745b8c/src/internal/util/reportUnhandledError.ts#L13-L14

While RxJS is by far the most popular stream library we don't specify that XState only works with that stream library. I've adjusted this test to showcase how this still crashes with other observable implementations and also added a second one that doesn't pass even with RxJS. That's because `removeChild` throws and we never actually get to sending the done event to a service.

```suggestion
it('should not crash on immediate spawned observable completion', () => {
    const createEmptyObservable = (): any => ({
      subscribe(_next, _error, complete) {
        complete();
      }
    });
    const parentMachine = createMachine<{
      child: ActorRef<any> | null;
    }>({
      context: {
        child: null
      },
      entry: assign({
        child: () => spawn(createEmptyObservable())
      })
    });
    const service = interpret(parentMachine);
    expect(() => {
      service.start();
    }).not.toThrow();
  });

  it('should receive done event from an immediately completed observable when self-initializing', () => {
    const parentMachine = createMachine<{
      child: ActorRef<any> | null;
    }>({
      context: {
        child: null
      },
      entry: assign({
        child: () => spawn(EMPTY, 'myactor')
      }),
      initial: 'init',
      states: {
        init: {
          on: {
            'done.invoke.myactor': 'done'
          }
        },
        done: {}
      }
    });
    const service = interpret(parentMachine);

    service.start();

    expect(service.state.value).toBe('done');
  });
```